app/configurator/configurator-state.spec.tsx

```diff
@@ -362,3 +379,85 @@ describe("moveField", () => {
     ]);
   });
 });
+
+describe("retainChartConfigWhenSwitchingChartType", () => {
+  const dataSetMetadata = covid19Metadata.data
+    .dataCubeByIri as DataCubeMetadata;
+  let oldConfig: ChartConfig;
+  let newConfig: ChartConfig;
+
+  const deriveNewChartConfig = (
+    oldConfig: ChartConfig,
+    newChartType: ChartType
+  ) => {
+    const newConfig = createDraft(
+      getChartConfigAdjustedToChartType({
+        chartConfig: oldConfig,
+        chartType: newChartType,
+        dimensions: dataSetMetadata.dimensions,
+        measures: dataSetMetadata.measures,
+      })
+    );
+    deriveFiltersFromFields(newConfig, dataSetMetadata);
+
+    return current(newConfig);
+  };
+
+  it("should retain appropriate chart config fields and discard the others", () => {
+    oldConfig = covid19ChartConfig as ColumnConfig;
+    newConfig = deriveNewChartConfig(oldConfig, "pie") as PieConfig;
+
+    // It should retain Y component iri between column and pie chart.
+    expect(newConfig.fields.y.componentIri).toEqual(
+      oldConfig.fields.y.componentIri
+    );
+
+    oldConfig = newConfig;
+    newConfig = deriveNewChartConfig(oldConfig, "column") as ColumnConfig;
+
+    // It should retain segment component iri when switching between charts
+    // that use segment field.
+    expect(newConfig.fields.segment?.componentIri).toEqual(
+      oldConfig.fields.segment.componentIri
+    );
+
+    oldConfig = newConfig;
+    newConfig = deriveNewChartConfig(oldConfig, "map") as MapConfig;
+
+    const expectedMapXComponentIri = getGeoShapesDimensions(
+      dataSetMetadata.dimensions
+    )[0].iri;
+
+    // It should not retain TemporalDimension on the X axis when switching to a
+    // map chart.
+    expect(newConfig.fields.areaLayer.componentIri).toEqual(
+      expectedMapXComponentIri
+    );
+    expect(newConfig.fields.areaLayer.componentIri).not.toEqual(
+      oldConfig.fields.x.componentIri
+    );
+
+    oldConfig = newConfig;
+    newConfig = deriveNewChartConfig(oldConfig, "column") as ColumnConfig;
+
+    // It should retain GeoShapeDimension on the X axis for a column chart.
+    expect(newConfig.fields.x.componentIri).toEqual(
+      oldConfig.fields.areaLayer.componentIri
+    );
+
+    oldConfig = newConfig;
+    newConfig = deriveNewChartConfig(oldConfig, "line") as LineConfig;
+
+    const expectedLineXComponentIri = findPreferredDimension(
+      dataSetMetadata.dimensions,
+      "TemporalDimension"
+    ).iri;
+
+    // It should not retain GeoShapesDimension on the X axis when switching to
+    // a line chart.
+    expect(newConfig.fields.x.componentIri).toEqual(expectedLineXComponentIri);
+    expect(newConfig.fields.x.componentIri).not.toEqual(
+      oldConfig.fields.x.componentIri
+    );
+  });
+});
```

Comment:
Is this the right way to test it? I wanted to mimic actual user behavior of switching chart types in a chain, but maybe this should be more explicit with actual chart config values visible to the user, e.g.

`const expectedChartConfig = {...}`

?

Reply:
I think it's OK right now.

One possibility I see would be to have it more declarative when you say what should happen.
Then you can build multiple scenarios more easily.

```
const checks = [
  {
    type: 'pie',
    checks: [
      ['y.componentIri', 'y.componentIri', true] 
    ]
  },
  { 
    type: 'column',
    checks: [['segment.componentIri', 'segment.componentIri', true]]
},
{
  type: 'map',
  checks:[
    ['areaLayer.componentIri', 'x.componentIri', false]
  ]
}
   
for (let nextChartType of nextChartTypes) {
   ... // code the expect etc :)
}

```

Reply:
Great idea, done, thanks! 💯